File: packages/go/fayda/userinfo.go
package fayda

import (
	"context"
	"fmt"
	"io"
	"net/http"

	"github.com/golang-jwt/jwt/v5"
)

// UserInfoModule handles userinfo fetching
type UserInfoModule struct {
	config *Config
	http   *http.Client
	auth   *AuthModule
	token  *TokenModule
}

// Get fetches the user's information using an access token
func (u *UserInfoModule) Get(ctx context.Context, accessToken string) (*FaydaUser, error) {
	if u.config.Sandbox {
		return MockUser(), nil
	}

	req, err := http.NewRequestWithContext(ctx, "GET", u.config.UserInfoEndpoint(), nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))

	resp, err := u.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &UserInfoError{message: fmt.Sprintf("userinfo request failed: %d", resp.StatusCode)}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// The response is a JWT, decode it without verification
	token, _, err := jwt.NewParser().ParseUnverified(string(body), jwt.MapClaims{})
	if err != nil {
		return nil, err
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return nil, fmt.Errorf("invalid claims format")
	}

	return mapClaimsToUser(claims)
}

// GetFromCode gets the user's information from an authorization code
func (u *UserInfoModule) GetFromCode(ctx context.Context, params GetFromCodeParams) (*FaydaUser, error) {
	if err := u.auth.ValidateCallback(params.State, params.ExpectedState); err != nil {
		return nil, err
	}

	tokens, err := u.token.Exchange(ctx, params.Code, params.CodeVerifier)
	if err != nil {
		return nil, err
	}

	return u.Get(ctx, tokens.AccessToken)
}

// GetMock returns a mock user (sandbox mode only)
func (u *UserInfoModule) GetMock() (*FaydaUser, error) {
	if !u.config.Sandbox {
		return nil, &SandboxError{message: "GetMock() is only available in sandbox mode"}
	}
	return MockUser(), nil
}

// mapClaimsToUser maps JWT claims to FaydaUser
func mapClaimsToUser(claims jwt.MapClaims) (*FaydaUser, error) {
	user := &FaydaUser{}

	// Map standard claims
	if sub, ok := claims["sub"].(string); ok {
		user.Sub = sub
	}
	if name, ok := claims["name"].(string); ok {
		user.Name = name
	}
	if nameAm, ok := claims["name#am"].(string); ok {
		user.NameAm = nameAm
	}
	if nameEn, ok := claims["name#en"].(string); ok {
		user.NameEn = nameEn
	}
	if email, ok := claims["email"].(string); ok {
		user.Email = email
	}
	if phone, ok := claims["phone_number"].(string); ok {
		user.PhoneNumber = phone
	}
	if picture, ok := claims["picture"].(string); ok {
		user.Picture = picture
	}
	if gender, ok := claims["gender"].(string); ok {
		user.Gender = gender
	}
	if birthdate, ok := claims["birthdate"].(string); ok {
		user.Birthdate = birthdate
	}

	// Map address
	if addressObj, ok := claims["address"].(map[string]interface{}); ok {
		addr := &FaydaAddress{}
		if formatted, ok := addressObj["formatted"].(string); ok {
			addr.Formatted = formatted
		}
		if locality, ok := addressObj["locality"].(string); ok {
			addr.Locality = locality
		}
		if country, ok := addressObj["country"].(string); ok {
			addr.Country = country
		}
		user.Address = addr
	}

	return user, nil
}

File: packages/rust/src/userinfo.rs
use crate::{auth::Auth, config::FaydaConfig, errors::FaydaError, models::*, token::Token};
use base64::Engine;

pub struct UserInfo {
    config: FaydaConfig,
}

impl UserInfo {
    pub fn new(config: FaydaConfig) -> Self {
        UserInfo { config }
    }

    /// Fetch user information using an access token
    pub async fn get(&self, access_token: &str) -> Result<FaydaUser, FaydaError> {
        if self.config.sandbox {
            return Ok(mock_user());
        }

        let client = reqwest::Client::new();
        let response = client
            .get(self.config.userinfo_endpoint())
            .header("Authorization", format!("Bearer {}", access_token))
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(FaydaError::UserInfo(format!(
                "UserInfo request failed: {}",
                response.status()
            )));
        }

        // Response is a signed JWT — decode payload without signature verification
        let jwt_text = response.text().await?;
        let claims = decode_jwt_payload(&jwt_text)?;
        Ok(map_claims_to_user(&claims))
    }

    /// Convenience method: validate callback → exchange code → fetch user
    pub async fn get_from_code(
        &self,
        params: GetFromCodeParams,
        auth: &Auth,
        token: &Token,
    ) -> Result<FaydaUser, FaydaError> {
        auth.validate_callback(&params.state, &params.expected_state)?;
        let tokens = token.exchange(&params.code, &params.code_verifier).await?;
        self.get(&tokens.access_token).await
    }

    /// Return mock user (sandbox mode only)
    pub fn get_mock(&self) -> Result<FaydaUser, FaydaError> {
        if !self.config.sandbox {
            return Err(FaydaError::Sandbox(
                "get_mock() is only available in sandbox mode".into(),
            ));
        }
        Ok(mock_user())
    }
}

fn decode_jwt_payload(jwt: &str) -> Result<serde_json::Value, FaydaError> {
    let parts: Vec<&str> = jwt.split('.').collect();
    if parts.len() != 3 {
        return Err(FaydaError::UserInfo("invalid JWT format".into()));
    }
    let payload_bytes = base64::engine::general_purpose::URL_SAFE_NO_PAD
        .decode(parts[1])
        .map_err(|e| FaydaError::UserInfo(format!("JWT base64 decode error: {}", e)))?;
    let claims: serde_json::Value = serde_json::from_slice(&payload_bytes)?;
    Ok(claims)
}

fn map_claims_to_user(claims: &serde_json::Value) -> FaydaUser {
    let address = claims.get("address").map(|a| FaydaAddress {
        formatted: a
            .get("formatted")
            .and_then(|v| v.as_str())
            .map(String::from),
        street_address: a
            .get("street_address")
            .and_then(|v| v.as_str())
            .map(String::from),
        locality: a.get("locality").and_then(|v| v.as_str()).map(String::from),
        region: a.get("region").and_then(|v| v.as_str()).map(String::from),
        country: a.get("country").and_then(|v| v.as_str()).map(String::from),
    });

    FaydaUser {
        sub: claims["sub"].as_str().unwrap_or("").to_string(),
        name: claims
            .get("name")
            .and_then(|v| v.as_str())
            .map(String::from),
        name_am: claims
            .get("name#am")
            .and_then(|v| v.as_str())
            .map(String::from),
        name_en: claims
            .get("name#en")
            .and_then(|v| v.as_str())
            .map(String::from),
        email: claims
            .get("email")
            .and_then(|v| v.as_str())
            .map(String::from),
        phone_number: claims
            .get("phone_number")
            .and_then(|v| v.as_str())
            .map(String::from),
        picture: claims
            .get("picture")
            .and_then(|v| v.as_str())
            .map(String::from),
        gender: claims
            .get("gender")
            .and_then(|v| v.as_str())
            .map(String::from),
        birthdate: claims
            .get("birthdate")
            .and_then(|v| v.as_str())
            .map(String::from),
        address,
    }
}

fn mock_user() -> FaydaUser {
    FaydaUser {
        sub: "mock-fayda-id-000001".to_string(),
        name: Some("Abebe Bikila".to_string()),
        name_am: Some("አበበ ቢኪላ".to_string()),
        name_en: Some("Abebe Bikila".to_string()),
        email: Some("abebe.bikila@example.com".to_string()),
        phone_number: Some("+251911000000".to_string()),
        picture: Some("https://placehold.co/150x150?text=AB".to_string()),
        gender: Some("male".to_string()),
        birthdate: Some("1932-08-07".to_string()),
        address: Some(FaydaAddress {
            formatted: Some("Addis Ababa, Ethiopia".to_string()),
            locality: Some("Addis Ababa".to_string()),
            country: Some("ET".to_string()),
            ..Default::default()
        }),
    }
}
